openapi3/internalize_refs.go

```diff
@@ -0,0 +1,389 @@
+			}
+			return s.Ref
+		}
+		break
+	}
+	if doc.Components.Schemas == nil {
+		doc.Components.Schemas = make(Schemas)
+	}
+	doc.Components.Schemas[name] = s.Value.NewRef()
+	s.Ref = "#/components/schemas/" + name
+	return s.Ref
+}
+
+func (doc *T) addParameterToSpec(p *ParameterRef, refNameResolver RefNameResolver) string {
+	if p == nil || !isExternalRef(p.Ref) {
+		return ""
+	}
+	name := refNameResolver(p.Ref)
+	if _, ok := doc.Components.Parameters[name]; ok {
+		p.Ref = "#/components/parameters/" + name
+		return p.Ref
+	}
+
+	if doc.Components.Parameters == nil {
+		doc.Components.Parameters = make(ParametersMap)
+	}
+	doc.Components.Parameters[name] = &ParameterRef{Value: p.Value}
+	p.Ref = "#/components/parameters/" + name
+	return p.Ref
+}
+
+func (doc *T) addHeaderToSpec(h *HeaderRef, refNameResolver RefNameResolver) string {
+	if h == nil || !isExternalRef(h.Ref) {
+		return ""
+	}
+	name := refNameResolver(h.Ref)
+	if _, ok := doc.Components.Headers[name]; ok {
+		h.Ref = "#/components/headers/" + name
+		return h.Ref
+	}
+	if doc.Components.Headers == nil {
+		doc.Components.Headers = make(Headers)
+	}
+	doc.Components.Headers[name] = &HeaderRef{Value: h.Value}
+	h.Ref = "#/components/headers/" + name
+	return h.Ref
+}
+
+func (doc *T) addRequestBodyToSpec(r *RequestBodyRef, refNameResolver RefNameResolver) string {
+	if r == nil || !isExternalRef(r.Ref) {
+		return ""
+	}
+	name := refNameResolver(r.Ref)
+	if _, ok := doc.Components.RequestBodies[name]; ok {
+		r.Ref = "#/components/requestBodies/" + name
+		return r.Ref
+	}
+	if doc.Components.RequestBodies == nil {
+		doc.Components.RequestBodies = make(RequestBodies)
+	}
+	doc.Components.RequestBodies[name] = &RequestBodyRef{Value: r.Value}
+	r.Ref = "#/components/requestBodies/" + name
+	return r.Ref
+}
+
+func (doc *T) addResponseToSpec(r *ResponseRef, refNameResolver RefNameResolver) string {
+	if r == nil || !isExternalRef(r.Ref) {
+		return ""
+	}
+	name := refNameResolver(r.Ref)
+	if _, ok := doc.Components.Responses[name]; ok {
+		r.Ref = "#/components/responses/" + name
+		return r.Ref
+	}
+	if doc.Components.Responses == nil {
+		doc.Components.Responses = make(Responses)
+	}
+	doc.Components.Responses[name] = &ResponseRef{Value: r.Value}
+	r.Ref = "#/components/responses/" + name
+	return r.Ref
+}
+
+func (doc *T) addSecuritySchemeToSpec(ss *SecuritySchemeRef, refNameResolver RefNameResolver) string {
+	if ss == nil || !isExternalRef(ss.Ref) {
+		return ""
+	}
+	name := refNameResolver(ss.Ref)
+	if _, ok := doc.Components.SecuritySchemes[name]; ok {
+		ss.Ref = "#/components/securitySchemes/" + name
+		return ss.Ref
+	}
+	if doc.Components.SecuritySchemes == nil {
+		doc.Components.SecuritySchemes = make(SecuritySchemes)
+	}
+	doc.Components.SecuritySchemes[name] = &SecuritySchemeRef{Value: ss.Value}
+	ss.Ref = "#/components/securitySchemes/" + name
+	return ss.Ref
+}
+
+func (doc *T) addExampleToSpec(e *ExampleRef, refNameResolver RefNameResolver) string {
```

Comment:
```suggestion
func (doc *T) addExampleToSpec(e *ExampleRef, refNameResolver RefNameResolver) {
```
No call to this function actually use the returned value so let's drop that.